indexing-service/src/main/java/io/druid/indexing/common/task/CompactionTask.java

```diff
@@ -0,0 +1,301 @@
+                null, // null means all unique metrics
+                injector,
+                indexIO
+            ),
+            false
+        ),
+        tuningConfig
+    );
+  }
+
+  private static Pair<Map<DataSegment, File>, List<TimelineObjectHolder<String, DataSegment>>> prepareSegments(
+      TaskToolbox toolbox,
+      String dataSource,
+      Interval interval
+  ) throws IOException, SegmentLoadingException
+  {
+    final List<DataSegment> usedSegments = toolbox
+        .getTaskActionClient()
+        .submit(new SegmentListUsedAction(dataSource, interval, null));
+    final Map<DataSegment, File> segmentFileMap = toolbox.fetchSegments(usedSegments);
+    final List<TimelineObjectHolder<String, DataSegment>> timelineSegments = VersionedIntervalTimeline
+        .forSegments(usedSegments)
+        .lookup(interval);
+    return Pair.of(segmentFileMap, timelineSegments);
+  }
+
+  private static DataSchema createDataSchema(
+      String dataSource,
+      Interval interval,
+      DimensionsSpec dimensionsSpec,
+      IndexIO indexIO,
+      ObjectMapper jsonMapper,
+      List<TimelineObjectHolder<String, DataSegment>> timelineSegments,
+      Map<DataSegment, File> segmentFileMap
+  )
+      throws IOException, SegmentLoadingException
+  {
+    // find metadata for interval
+    final List<QueryableIndex> segments = loadSegments(timelineSegments, segmentFileMap, indexIO);
+
+    // set rollup only if rollup is set for all segments
+    final boolean rollup = segments.stream().allMatch(index -> index.getMetadata().isRollup());
+
+    // find merged aggregators
+    final List<AggregatorFactory[]> aggregatorFactories = segments.stream()
+                                                                  .map(index -> index.getMetadata().getAggregators())
+                                                                  .collect(Collectors.toList());
+    final AggregatorFactory[] mergedAggregators = AggregatorFactory.mergeAggregators(aggregatorFactories);
+
+    if (mergedAggregators == null) {
+      throw new ISE("Failed to merge aggregators[%s]", aggregatorFactories);
+    }
+
+    // conver to combining aggregators
+    final List<AggregatorFactory> combiningAggregatorList = Arrays.stream(mergedAggregators)
+                                                                  .map(AggregatorFactory::getCombiningFactory)
+                                                                  .collect(Collectors.toList());
+    final AggregatorFactory[] combiningAggregators = combiningAggregatorList.toArray(
+        new AggregatorFactory[combiningAggregatorList.size()]
+    );
+
+    // find granularity spec
+    final GranularitySpec granularitySpec = new ArbitraryGranularitySpec(
+        new NoneGranularity(),
+        rollup,
+        ImmutableList.of(interval)
+    );
+
+    return new DataSchema(
+        dataSource,
+        dimensionsSpec == null ? ImmutableMap.of("type", "noop")
```

Comment:
It would be more type-safe to do something like `jsonMapper.convertValue(parser, JacksonUtils.TYPE_REFERENCE_MAP_STRING_OBJECT)`, on a parser you create using a normal constructor.

Reply:
Thanks. Changed.